First functions related to class

maxence-charriere opened the pull request:
@achille-roussel
Hey Ach tu pourrai jeter un oeil la dessus et me faire un retour?
Pour les nom de fiction je n’ai pas utilise de underscore.
J’utilise golint, il me dit que c’est pas bien.

voila le lien que j’utilise pour calquer:
https://developer.apple.com/library/mac/documentation/Cocoa/Reference/Ob
jCRuntimeRef/

maxence-charriere committed:
First functions related to class

@achille-roussel
Hey Ach tu pourrai jeter un oeil la dessus et me faire un retour?
Pour les nom de fiction je n’ai pas utilise de underscore.
J’utilise golint, il me dit que c’est pas bien.

voila le lien que j’utilise pour calquer:
https://developer.apple.com/library/mac/documentation/Cocoa/Reference/Ob
jCRuntimeRef/

achille-roussel commented:
Je pense qu'il faut dire à golint d'aller se faire voir il capte pas ce qui se passe ici, le risque est faible mais tu pourrais te retrouver avec différentes fonctions C:

``` c
void classGetName();

void class_get_name();

void class_getName();
```

qui auraient le même nom en Go du coup, `ClassGetName`... Je pense vraiment que coller au max à l'API initiale c'est le plus important ici, ça permet à n'importe qui qui est familier avec Objective-C d'être à l'aise très vite sans avoir besoin de réfléchir pour chaque fonction utilisée quel nom elle aura en Go.

achille-roussel commented on cgo.go:
```diff
@@ -0,0 +1,5 @@
+package objcruntime
```
Pourquoi pas appeler ça juste `objc` ?

Il va falloir répéter `objcruntime.` pour chaque utilisation de symboles du package... ça me semble lourd et pas trop d'en l'esprit du langage: https://golang.org/doc/effective_go.html#package-names

achille-roussel commented on class.go:
```diff
@@ -0,0 +1,94 @@
+// ClassGetSuperclass returns the superclass of a class.
+//
+// Discussion:
+// - You should usually use NSObject‘s superclass method instead of this function.
```
`Copyright © 2015 Apple Inc. All rights reserved.` C'est assez risqué de copier la doc d'Apple ;)

achille-roussel commented on class.go:
```diff
@@ -0,0 +1,94 @@
+//
+// Discussion:
+// - You should usually use NSObject‘s superclass method instead of this function.
+func ClassGetSuperclass() (cls Class) {
```
T'as pas besoin de nommer la valeur de retour ici, elle est jamais utilisée (golint fail...)

maxence-charriere replied:
Petite erreur quand j'ai refractor.

achille-roussel commented on class.go:
```diff
@@ -0,0 +1,94 @@
+
+// ClassGetInstanceVariable returns the Ivar for a specified instance variable of a given class.
+func ClassGetInstanceVariable(cls Class, name string) Ivar {
+	cname := C.CString(string(name))
```
Pourquoi est-ce que tu répètes `string(name)`, la variable est déjà de type `string` ?

achille-roussel commented on class.go:
```diff
@@ -0,0 +1,94 @@
+}
+
+// ClassCopyIvarList describes the instance variables declared by a class.
+func ClassCopyIvarList(cls Class) (ivars []Ivar, outCount uint) {
```
Je plussois l'utilisation de plusieurs valeurs de retour au lieu du passage de pointeur :+1:

achille-roussel commented on class.go:
```diff
@@ -0,0 +1,94 @@
+// ClassGetClassVariable returns the Ivar for a specified class variable of a given class.
+func ClassGetClassVariable(cls Class, name string) Ivar {
+	cname := C.CString(string(name))
+	defer C.free(unsafe.Pointer(cname))
```
Toutes ces allocations dynamiques de c-string ça peut potentiellement impacter les perf, je pense que ça sera intéressant d'optimiser tout ça le moment venu ;)

achille-roussel commented on class.go:
```diff
@@ -0,0 +1,94 @@
+
+	outCount = uint(coutCount)
+
+	if outCount > 0 {
```
Embrasse le pouvoir de Go mec:

``` go
if outCount = uint(coutCount); outCount > 0 {
    ...
}
```

achille-roussel commented on class.go:
```diff
@@ -0,0 +1,94 @@
+	outCount = uint(coutCount)
+
+	if outCount > 0 {
+		ivars = make([]Ivar, 0, outCount)
```
Quand tu connais la taille de ta slice, plutôt que d'utiliser `append` en boucle tu peux utiliser ce pattern qui est plus efficient:

``` go
ivars = make([]Ivar, outCount)

for i := uint(0); i != outCount; i++ {
    ...
    ivars[int(i)] = ...
}
```

achille-roussel commented:
Au niveau de la structure du projet, tu vas gagner en temps de compilation si tu limite le nombre d'inclusions de header C, plutôt que d'avoir `class.go`, `ivar.go`, etc... qui vont tous inclure `objc/runtime.h` tu pourrais avoir un seul `runtime.go` qui défini les bindings de tous les types et fonctions du header.
(le coût est linéaire, si il faut 1 seconde pour compiler un fichier, chaque nouveau fichier va ajouter 1 seconde).

achille-roussel commented:
Il manque quelques fonctions aussi, check `/usr/include/objc/runtime.h` ;)

maxence-charriere commented:
Oula il n'en manque pas que quelques ^^.

maxence-charriere committed:
refractor style

maxence-charriere commented:
@achille-roussel  
Modifé suivant tes conseils.
Pour les fichiers je rassemblerai tout plus tard, pour le moment ca m'aide d'avoir les choses séparés par theme.
Je continuerai tout ca pendant que Pouis regarde pas :D

achille-roussel commented on class.go:
```diff
@@ -0,0 +1,71 @@
+
+type Class C.Class
+
+func Class_GetName(cls Class) string {
```
Pour coller à l'API je pense que ces fonctions devraient être `Class_getName`, `Class_getSuperClass`, ...

achille-roussel commented on class.go:
```diff
@@ -0,0 +1,71 @@
+	return C.class_addIvar(cls, cname, C.size_t(size), C.uint8_t(alignment), ctypes) == 1
+}
+
+func Class_CopyIvarList(cls Class) (ivarList []Ivar, outCount uint) {
```
Au final je pense que c'est pas nécessaire de renvoyer `outCount` ici, ça aura toujours la même valeur que `len(ivarList)`.

maxence-charriere committed:
class refract

maxence-charriere commented:
@achille-roussel ok

achille-roussel commented on class.go:
```diff
@@ -0,0 +1,71 @@
+	ctypes := C.CString(string(types))
+	defer C.free(unsafe.Pointer(ctypes))
+
+	return C.class_addIvar(cls, cname, C.size_t(size), C.uint8_t(alignment), ctypes) == 1
```
Pour ça tu devrais comparer `!= 0`, la seule valeur qui soit `false` en C c'est zero, tout le reste est `true`.

achille-roussel commented on class.go:
```diff
@@ -0,0 +1,71 @@
+}
+
+func Class_getClassVariable(cls Class, name string) Ivar {
+	cname := C.CString(string(name))
```
T'as encore quelques `string(x)` inutiles dans ton code.

maxence-charriere committed:
adds

achille-roussel commented on class.go:
```diff
@@ -0,0 +1,130 @@
+}
+
+func Class_isMetaClass(cls Class) bool {
+	return C.class_isMetaClass(cls) == 1
```
`!= 0` ici aussi.

achille-roussel replied:
Au passage, les CPUs ont des instructions spéciales pours les comparaisons avec zéro, le compilo peut générer un meilleur code si tu compares à zéro plutôt qu'à une autre valeur.

achille-roussel commented on class.go:
```diff
@@ -0,0 +1,130 @@
+	clayout := unsafe.Pointer(C.CString(layout))
+	defer C.free(clayout)
+
+	C.class_setIvarLayout(cls, (*C.uint8_t)(clayout))
```
Les sources du runtime Objective-C sont ouvertes, il faudrait vérifier si le `layout` est copié en interne dans la classe. Si c'est pas le cas ce code va crashé au bout d'un moment parce que l'espace mémoire est libéré à la sortie de la fonction.

achille-roussel replied:
Aussi je pense que c'est pas des chaines de caractères (au sens ASCII), regarde ça http://stackoverflow.com/questions/16131172/what-are-class-setivarlayout-and-class-getivarlayout

Les layouts sont juste des pointeurs, je pense qu'il faudrait changer la définition de ces fonctions comme ça:

``` go
func Class_getIvarLayout(cls Class) uintptr {
    return uintptr(unsafe.Pointer(C.class_getIvarLayout(cls)))
}

func Class_setIvarLayout(cls Class, layout uintptr) {
    C.class_setIvarLayout(cls, (*C.uint8_t)(unsafe.Pointer(layout)))
}
```

maxence-charriere committed:
refract bis

achille-roussel commented on class.go:
```diff
@@ -0,0 +1,126 @@
+	return C.GoString((*C.char)(clayout))
+}
+
+func Class_setWeakIvarLayout(cls Class, layout string) {
```
Pareil que pour `Class_setIvarLayout` ;)

achille-roussel commented on class.go:
```diff
@@ -0,0 +1,126 @@
+//
+// static objc_property_t *property_offset(objc_property_t *p, size_t n) {
+//   return p + n;
+// }
```
J'ai fais quelques recherches et le coût pour appeler une fonction C depuis Go est ~10x le coût d'un appel à une fonction Go simple.

J'ai abandonné ce genre d'approche parce que c'est du code qui est typiquement appelé en boucle donc l'impact est d'autant plus grand.

À la place j'ai utilisé une fonction comme ça:

``` go
func next(list *C.list) *C.list {
    return (*C.list)(unsafe.Pointer(uintptr(list) + unsafe.Sizeof(*list)))
}
```

maxence-charriere committed:
refractor + selector

achille-roussel commented on runtime.go:
```diff
@@ -0,0 +1,206 @@
+	return Method(C.class_getInstanceMethod(aClass, aSelector))
+}
+
+func Class_getClassMethodt(aClass Class, aSelector Sel) Method {
```
t'as un `t` en trop à la fin ;)

achille-roussel commented on runtime.go:
```diff
@@ -0,0 +1,206 @@
+//
+// static Method *method_offset(Method *p, size_t n) {
+//   return p + n;
+// }
```
Je crois que t'as plus besoin de ce bout de code.

achille-roussel commented on runtime_test.go:
```diff
@@ -0,0 +1,17 @@
+	if selBName := Sel_getName(selB); selAName != selBName {
+		t.Errorf("selAName and selBName are not equal: %s != %s\n", selAName, selBName)
+	}
+}
```
Dans l'esprit des tests unitaires tu devrais séparer ce test un deux je pense, un où tu testes `Sel_isEqual` et un autre où tu testes  que `Sel_getName` renvoie bien la même valeur.

L'idée c'est que si un teste retourne une erreur tu puisses identifier directement ce qui ne va pas parce que le teste vérifie une seule fonction.

maxence-charriere committed:
refractor + add

achille-roussel commented on runtime_test.go:
```diff
@@ -0,0 +1,35 @@
+
+	if Sel_isEqual(selA, selB) {
+		t.Errorf("selA: %p and selB: %p should not be equal", selA, selB)
+	}
```
Ça je pense que tu peux le séparer en deux tests aussi, un pour tester l'égalité et un pour tester la différence.

achille-roussel commented on runtime_test.go:
```diff
@@ -0,0 +1,35 @@
+
+	if selAName == selBName {
+		t.Errorf("selAName and selBName should not be equal: %s == %s\n", selAName, selBName)
+	}
```
Pareil ici, un test pour tester l'égalité des noms, un autre pour tester la différence.

maxence-charriere committed:
selector test + property

@achille-roussel  j’ai remis plusieurs fichier. J’aimerai garder ce
decoupage le temps de l’implementation.
Les test des propriété viendront quand je pourrai créer une classe.
Je serai fixé quant au free a faire sur PropertyAttr quand j’aurai mes
tests.

maxence-charriere committed:
class first section + some protocol

maxence-charriere committed:
protocol bindings

@achille-roussel Tests is coming!
Mais la Poui commence a tirer la tronche ^^

achille-roussel commented on coverage.out:
```diff
@@ -0,0 +1,75 @@
+github.com/maxence-charriere/go-objcruntime/class.go:256.43,258.2 1 0
+github.com/maxence-charriere/go-objcruntime/class.go:260.52,262.2 1 0
+github.com/maxence-charriere/go-objcruntime/class.go:265.47,268.2 2 0
+github.com/maxence-charriere/go-objcruntime/class.go:270.53,273.2 2 0
```
I think you shouldn't commit this file, add it to the `.gitignore` so it doesnt' get added automatically.

achille-roussel commented on property.go:
```diff
@@ -0,0 +1,80 @@
+	Value string
+}
+
+func (attr PropertyAttribute) ctype() C.objc_property_attribute_t {
```
Ces fonctions `ctype` sont pas utilisées, c'est normal ?

maxence-charriere replied:
yep je les avait utilisé et j ai fait un refractor or j en avait plus besoin. Juste oublié de les virer

maxence-charriere committed:
clean up

maxence-charriere committed:
cleanup

maxence-charriere committed:
git ignore

achille-roussel commented on protocol.go:
```diff
@@ -0,0 +1,173 @@
+		elem.value = C.CString(attr.Value)
+
+		defer C.free(unsafe.Pointer(elem.name))
+		defer C.free(unsafe.Pointer(elem.value))
```
Il y a peut de chances que se soit un problème ici mais tu devrais suivre ce pattern autant que possible:

``` go
elem.name = C.CString(attr.Name)
defer C.free(unsafe.Pointer(elem.name))

elem.value = C.CString(attr.Value)
defer C.free(unsafe.Pointer(elem.value))
```

Définir le `defer` pour libérer la mémoire juste après l'allocation pourra éviter un fuite mémoire si une des operations entre les deux panique.

achille-roussel replied:
Au passage je pense que dans ce cas il y a une approche plus intéressante, `defer` va consommer le la mémoire (sur la _stack_ si possible, des fois sur la _heap_).
Faire un `defer` en boucle va augmenter l'utilisation mémoire de cet appel à fonction de façon linéaire avec le nombre d'éléments dans la liste, ce qui peut causer des problèmes.

Une autre approche peut être celle ci:

``` go
cattributes := (*C.objc_property_attribute_t)(C.calloc(C.size_t(attributeCount), C.size_t(attrSize)))

defer func(attr *C.objc_property_attribute_t, n int) {
    elem := attr
    for i := 0; i != n; i++ {
        C.free(unsafe.Pointer(elem.name))
        C.free(unsafe.Pointer(elem.value))
        elem = nextPropertyAttr(elem)
    }
    C.free(unsafe.Pointer(attr))
}(cattributes, attributeCount)
```

De cette façon tu n'utilises qu'un `defer` et l'utilisation mémoire de cette fonction devient constante.

maxence-charriere committed:
tests for protocols and properties

maxence-charriere committed:
Merge branch 'master' into class

maxence-charriere committed:
more tests + react

maxence-charriere committed:
more tests (coverage 100%)

maxence-charriere commented:
@achille-roussel Hey mec est ce que tu peux review ca, le test coverage est a 100% :sunglasses:

achille-roussel commented on class.go:
```diff
@@ -0,0 +1,267 @@
+		defer func(cattributes *C.objc_property_attribute_t, attributeCount int) {
+			elem := cattributes
+
+			for i := 0; i < attributeCount; i++ {
```
Tu peux mettre `elem` dans la boucle `for` puisque c'est pas utilisé en dehors:

``` go
for i, elem := 0, cattributes; ... {

}
```

maxence-charriere replied:
ohhhhh ok ;)

achille-roussel commented on class_test.go:
```diff
@@ -0,0 +1,557 @@
+	nsObjectProto := Objc_getProtocol("NSObject")
+
+	if Class_conformsToProtocol(class, nsObjectProto) {
+		t.Error("class should not conform to NSOject protocol")
```
T'as écris `NSOject` plusieurs fois :P

achille-roussel commented:
Mec ça m'a l'air bien cool tout ça !

maxence-charriere commented:
je vais modifier avec tes remarques et faire le premier merge. la suite viendra ensuite :)

maxence-charriere committed:
refact

@achille-roussel

maxence-charriere committed:
react NSObject

maxence-charriere commented:
@achille-roussel Up pour merge ?

achille-roussel commented:
Go
